design/plugin-versioning.md

```diff
@@ -0,0 +1,232 @@
+- Allow for backwards and forwards compatibility with plugin versions
+- Establish a design process for adding new plugin API methods
+
+## Non Goals
+
+- Support breaking changes in the plugin APIs such as method removal or method signature changes
+- Changing how plugins are managed or added
+
+## High-Level Design
+
+With each change to a plugin API, a new version of the plugin interface and the proto service definition will be created which incorporates the new changes plus any previous versions of that particular plugin API.
+The plugin framework will be adapted to allow these new plugin versions to be registered.
+Plugins can opt to implement any or all versions of an API, however Velero will always attempt to use the latest version, and the plugin management will be modified to adapt earlier versions of a plugin to be compatible with the latest API.
+Under the existing plugin framework, any new plugin version will be treated as a new plugin with a new kind, however the plugin manager (which provides implementations of a plugin to Velero) will include an adapter layer which will manage the different versions and adapt any version which does not implement the latest version of the plugin API.
+
+For now, there will be a restriction that only new methods can be added to a plugin API.
+Methods can not be removed or modified.
+This restriction will be in place until Velero v2 when backwards incompatible changes may be included.
+
+Although adding new rpc methods to a service is considered a backwards compatible change within gRPC, due to the way the proto definitions are compiled and included in the framework used by plugins, this will require every plugin to implement the new methods.
+Instead, we are opting to treat the addition of a method to an API as one requiring versioning.
+
+The addition of optional fields to existing structs which are used as parameters to or return values of API methods will not be considered as a change requiring versioning.
+These kinds of changes do not modify method signatures and have been safely made in the past with no impact on existing plugins.
+
+## Detailed Design
+
+The following areas will need to be adapted to support plugin versioning.
+
+### Plugin Interface Definitions
+
+To provide versioned plugins, any new method added to a plugin interface will require a new versioned interface to be created which incorporates previous versions of the interface.
+Currently, all plugin interface definitions reside in `pkg/plugin/velero` in a file corresponding to their plugin kind.
+These files will be rearranged to be grouped by kind and then versioned: `pkg/plugin/velero/<plugin_kind>/<version>/`.
+
+For example, if a new method belongs to the ObjectStore API, a file would be added to `pkg/plugin/velero/objectstore/v2/` and would contain the following new API definition:
+
+```
+import "github.com/vmware-tanzu/velero/pkg/plugin/velero/objectstore/v1""
```

Comment:
If we are to allow break changes the new interface should not need to shadow the v1 interfaces.